plugins/outputs/sql/sql.go

```diff
@@ -0,0 +1,256 @@
+  ## or a simple string:
+  ##   host=localhost user=pqotest password=... sslmode=... dbname=app_production
+  ##
+  ## All connection parameters are optional.
+  ##
+  ## Without the dbname parameter, the driver will default to a database
+  ## with the same name as the user. This dbname is just for instantiating a
+  ## connection with the server and doesn't restrict the databases we are trying
+  ## to grab metrics for.
+  ##
+  address = "username:password@tcp(server:port)/table"
+
+  ## Available Variables:
+  ##   {TABLE} - tablename as identifier
+  ##   {TABLELITERAL} - tablename as string literal
+  ##   {COLUMNS} - column definitions
+  ##   {KEY_COLUMNS} - comma-separated list of key columns (time + tags)
+  ##
+
+  ## Check with this is table exists
+  ##
+  ## Template for MySQL is "SELECT 1 FROM {TABLE} LIMIT 1"
+  ##
+  table_exists_template = "SELECT 1 FROM {TABLE} LIMIT 1"
+
+  ## Template to use for generating tables
+
+  ## Default template
+  ##
+  # table_template = "CREATE TABLE {TABLE}({COLUMNS})"
+
+  ## Convert Telegraf datatypes to these types
+  #[outputs.sql.convert]
+  #  integer              = "INT"
+  #  real                 = "DOUBLE"
+  #  text                 = "TEXT"
+  #  timestamp            = "TIMESTAMP"
+  #  defaultvalue         = "TEXT"
+  #  unsigned             = "UNSIGNED"
+`
+
+func (p *Sql) SampleConfig() string { return sampleConfig }
+func (p *Sql) Description() string  { return "Send metrics to SQL Database" }
+
+func (p *Sql) generateCreateTable(metric telegraf.Metric) string {
+	var columns []string
+	var pk []string
+	var sql []string
+
+	pk = append(pk, quoteIdent("timestamp"))
+	columns = append(columns, fmt.Sprintf("timestamp %s", p.Convert.Timestamp))
+
+	// tags in measurement table
+	for _, tag := range metric.TagList() {
+		pk = append(pk, quoteIdent(tag.Key))
+		columns = append(columns, fmt.Sprintf("%s %s", quoteIdent(tag.Key), p.Convert.Text))
+	}
+
+	var datatype string
+	for _, field := range metric.FieldList() {
+		datatype = p.deriveDatatype(field.Value)
+		columns = append(columns, fmt.Sprintf("%s %s", quoteIdent(field.Key), datatype))
+	}
+
+	var query string
+	query = strings.Replace(p.TableTemplate, "{TABLE}", quoteIdent(metric.Name()), -1) //metric name
+	query = strings.Replace(query, "{TABLELITERAL}", quoteLiteral(metric.Name()), -1)  //quoted metric name
+	query = strings.Replace(query, "{COLUMNS}", strings.Join(columns, ","), -1)
+	query = strings.Replace(query, "{KEY_COLUMNS}", strings.Join(pk, ","), -1)
+
+	sql = append(sql, query)
+	return strings.Join(sql, ";")
+}
+
+func (p *Sql) generateInsert(tablename string, columns []string) string {
+
+	var placeholder, quoted []string
+	for _, column := range columns {
+		placeholder = append(placeholder, fmt.Sprintf("?"))
+		quoted = append(quoted, quoteIdent(column))
+	}
+
+	sql := fmt.Sprintf("INSERT INTO %s(%s) VALUES(%s)", quoteIdent(tablename), strings.Join(quoted, ","), strings.Join(placeholder, ","))
```

Comment:
column names might need to be escaped for some weirder column names.

Reply:
Yes, there's also opportunity for a sql injection attack through the names of metrics, fields, and tags currently. I'll have to escape them to get them to work and sanitize them so they can't be misused.

Reply:
I added identifier quoting and a check for invalid characters that should cover this.